src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs

```diff
@@ -0,0 +1,619 @@
+        private bool SameAsyncInterface(INamedTypeSymbol interfaceA, INamedTypeSymbol interfaceB)
+        {
+            /* Using OriginalDefinition b/c the generic field of the metadata type has the template name, e.g. `TProgress`
+             * and the actual interface will have a concrete type, e.g. `int` */
+            return SymbolEqualityComparer.Default.Equals(interfaceA.OriginalDefinition, interfaceB.OriginalDefinition);
+        }
+
+        /// <summary>
+        /// returns true if the class represented by the symbol implements any of the interfaces defined in ProhibitedAsyncInterfaces
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="classSymbol"></param>
+        /// <param name="classDeclaration"></param>
+        /// <returns>True iff the given class implements any of the IAsync interfaces that are not valid in Windows Runtime</returns>
+        public bool ImplementsAsyncInterface(ref GeneratorExecutionContext context, INamedTypeSymbol classSymbol, ClassDeclarationSyntax classDeclaration)
+        {
+            foreach (string prohibitedInterface in ProhibitedAsyncInterfaces)
+            {
+                INamedTypeSymbol asyncInterface = context.Compilation.GetTypeByMetadataName(prohibitedInterface);
+                foreach (INamedTypeSymbol interfaceImplemented in classSymbol.AllInterfaces)
+                {
+                    if (SameAsyncInterface(interfaceImplemented, asyncInterface))
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(DiagnosticRules.AsyncRule, classDeclaration.GetLocation(), classDeclaration.Identifier, interfaceImplemented));
+                        return true;
+                        /* By exiting early, we only report diagnostic for first prohibited interface we see. 
+                        If a class implemented 2 (or more) such interfaces, then we would only report diagnostic error for the first one. 
+                        could thread `found` variable from CatchWinRTDiagnostics here as well, if we want more diagnostics reported */
+                    }
+                }
+            }
+            return false;
+        }
+
+        #endregion
+
+        /// <summary>
+        /// keeps track of the arity of all constructors, and reports the diagnostic (and exits) as soon as a two constructors of the same arity are seen.
+        /// </summary>
+        /// <param name="context">Object to raise diagnostic on</param>
+        /// <param name="classDeclaration">look for constructors of this class</param>
+        /// <returns>True if multiple constructors of the same arity exist for the given class</returns>
+        public bool HasMultipleConstructorsOfSameArity(ref GeneratorExecutionContext context, ClassDeclarationSyntax classDeclaration)
+        {
+            IEnumerable<ConstructorDeclarationSyntax> constructors = classDeclaration.ChildNodes().OfType<ConstructorDeclarationSyntax>();

```

Comment:
Do we need a public check as I would assume if they were private, it might be fine.